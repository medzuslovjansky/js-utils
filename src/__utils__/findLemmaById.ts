import fs from 'node:fs';
import path from 'node:path';

const fixturesDir = path.join(__dirname, '../../src/__fixtures__');

let fixtures: Map<string, string[]> | undefined;

export function findLemmaById(id: string): string {
  return initFixtures().get(id)?.[2] ?? id;
}

function initFixtures() {
  if (!fixtures) {
    const map = (fixtures = new Map());
    fs.readdirSync(fixturesDir).forEach((name) => {
      const jsonPath = path.join(fixturesDir, name);
      const json = fs.readFileSync(jsonPath, 'utf-8');
      const data = JSON.parse(json);
      for (const row of data) {
        map.set(row[0], row);
      }
    });
  }

  return fixtures;
}
