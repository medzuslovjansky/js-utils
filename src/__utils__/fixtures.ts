type FixtureGetter = () => [string, string, string, string][];

export const adjectives: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/adjectives.json');
export const nouns_feminine: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-feminine.json');
export const nouns_masculine_animate: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-masculine-animate.json');
export const nouns_masculine: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-masculine.json');
export const nouns_misc: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-misc.json');
export const nouns_neuter: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/nouns-neuter.json');
export const verbs_imperfect: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-imperfect.json');
export const verbs_perfect: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-perfect.json');
export const verbs_misc: FixtureGetter = () =>
  jest.requireActual('../__fixtures__/verbs-misc.json');
