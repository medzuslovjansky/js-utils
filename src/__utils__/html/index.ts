export * as adjective from './adjective';
export * as noun from './noun';
export * as verb from './verb';
