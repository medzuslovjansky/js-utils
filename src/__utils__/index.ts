export * from './toMarkdown';
export * from './findLemmaById';
export * as fixtures from './fixtures';
