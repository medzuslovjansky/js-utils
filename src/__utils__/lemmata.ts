import fs from 'node:fs';
import path from 'node:path';
import { parsePos } from '../partOfSpeech';

const fixturesDir = path.join(__dirname, '../../src/__fixtures__');

let fixtures: Map<string, string[]> | undefined;

export function findLemmaById(id: string): string {
  return initFixtures().get(id)?.[2] ?? id;
}

export function findPosTags(id: string): string[] {
  const pos = initFixtures().get(id)?.[1];
  if (!pos) return [];

  const { name, ...attributes } = parsePos(pos);
  const trueAttributes = Object.entries(attributes)
    .filter(([_, value]) => value)
    .map(([key]) => key);

  return [name, ...trueAttributes];
}

function initFixtures() {
  if (!fixtures) {
    const map = (fixtures = new Map());
    fs.readdirSync(fixturesDir).forEach((name) => {
      const jsonPath = path.join(fixturesDir, name);
      const json = fs.readFileSync(jsonPath, 'utf-8');
      const data = JSON.parse(json);
      for (const row of data) {
        map.set(row[0], row);
      }
    });
  }

  return fixtures;
}
