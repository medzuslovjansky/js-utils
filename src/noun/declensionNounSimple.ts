import { parsePos } from '../partOfSpeech';
import { declensionNoun } from './declensionNoun';

export function declensionNounSimple(
  lemma: string,
  morphology: string,
  extra = '',
  genderOverride?: 'masculine' | 'feminine',
) {
  const pos = parsePos(morphology);
  if (pos.name !== 'noun') throw new Error('not a noun');

  let gender = pos.gender;
  if (gender === 'masculineOrFeminine') {
    gender = genderOverride ?? gender;
  }

  return declensionNoun(
    lemma,
    extra,
    gender,
    pos.animate,
    pos.plural,
    pos.singular,
    pos.indeclinable,
  );
}
