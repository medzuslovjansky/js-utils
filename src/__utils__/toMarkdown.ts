import vm from 'node:vm';
import type { TestCaseResult } from '@jest/reporters';
import { transliterate } from '..';
import { toMarkdown as verb } from '../verb/__utils__/toMarkdown';
import { SteenVerbParadigm } from '../verb';

type Tuple<T> = [T, T];
type TupleProcessor<T> = (diff: Tuple<T>) => string;

export function toMarkdown(test: TestCaseResult) {
  const processor = getMarkdownProcessor(test);
  const tables = processor ? getBeforeAndAfter(test).map(processor) : [];
  return transliterate(
    tables.map((table) => table + '\n\n').join(''),
    'art-Latn-x-interslv-etym',
  );
}

function getBeforeAndAfter(test: TestCaseResult): Tuple<unknown>[] {
  return (test.failureDetails as any[])
    .map((failure) => failure?.matcherResult)
    .filter(
      (matcherResult) =>
        matcherResult && matcherResult.expected && matcherResult.actual,
    )
    .map((failure) =>
      vm.runInNewContext(`[${failure.expected}, ${failure.actual}]`),
    );
}

function getMarkdownProcessor(
  test: TestCaseResult,
): TupleProcessor<unknown> | undefined {
  const partOfSpeech = test.ancestorTitles[0];
  switch (partOfSpeech) {
    case 'verb':
      return (tuple) => verb(tuple as Tuple<SteenVerbParadigm>);
    default:
      return;
  }
}
