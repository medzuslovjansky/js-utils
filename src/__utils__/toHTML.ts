import vm from 'node:vm';

import type { TestCaseResult } from '@jest/reporters';

import { SteenVerbParadigm } from '../verb';
import { SteenNounParadigm } from '../noun';
import { SteenAdjectiveParadigm } from '../adjective';

import * as html from './html';

type Tuple<T> = [T, T];
type TupleProcessor<T> = (diff: Tuple<T>) => string;

export function toHTML(test: TestCaseResult) {
  const renderer = getRenderer(test);
  const figures = renderer ? getBeforeAndAfter(test).map(renderer) : [];
  return figures.map((figure) => figure + '\n\n').join('');
}

function getBeforeAndAfter(test: TestCaseResult): Tuple<unknown>[] {
  return (test.failureDetails as any[])
    .map((failure) => failure?.matcherResult)
    .filter(
      (matcherResult) =>
        matcherResult && matcherResult.expected && matcherResult.actual,
    )
    .map((failure) =>
      vm.runInNewContext(`[${failure.expected}, ${failure.actual}]`),
    );
}

function getRenderer(
  test: TestCaseResult,
): TupleProcessor<unknown> | undefined {
  const partOfSpeech = test.ancestorTitles[0];
  switch (partOfSpeech) {
    case 'verb':
      return (tuple) =>
        html.verb.renderDiff(...(tuple as Tuple<SteenVerbParadigm>));
    case 'noun':
      return (tuple) =>
        html.noun.renderDiff(...(tuple as Tuple<SteenNounParadigm>));
    case 'adjective':
      return (tuple) =>
        html.adjective.renderDiff(...(tuple as Tuple<SteenAdjectiveParadigm>));
    default:
      return;
  }
}
