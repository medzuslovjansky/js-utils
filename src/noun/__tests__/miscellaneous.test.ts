import { declensionNounSimple } from '../declensionNounSimple';
import { nouns_misc } from '../../__utils__/fixtures';

describe('noun', () => {
  describe('miscellaneous', () => {
    test.each(nouns_misc())(
      '%s (as masculine)',
      (_id, morphology, lemma, extra) => {
        const masculine = declensionNounSimple(
          lemma,
          morphology,
          extra,
          'masculine',
        );
        expect(masculine).toMatchSnapshot('masculine');
      },
    );

    test.each(nouns_misc())(
      '%s (as feminine)',
      (_id, morphology, lemma, extra) => {
        const feminine = declensionNounSimple(
          lemma,
          morphology,
          extra,
          'feminine',
        );
        expect(feminine).toMatchSnapshot('feminine');
      },
    );
  });
});
