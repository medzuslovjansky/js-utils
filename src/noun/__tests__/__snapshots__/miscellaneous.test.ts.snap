// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`noun miscellaneous 339 (as feminine): feminine 1`] = `
{
  "acc": [
    "luč",
    "luči",
  ],
  "dat": [
    "luči",
    "lučam",
  ],
  "gen": [
    "luči",
    "lučij",
  ],
  "ins": [
    "lučjų",
    "lučami",
  ],
  "loc": [
    "luči",
    "lučah",
  ],
  "nom": [
    "luč",
    "luči",
  ],
  "voc": [
    "luči",
    "luči",
  ],
}
`;

exports[`noun miscellaneous 339 (as masculine): masculine 1`] = `
{
  "acc": [
    "luč",
    "luče",
  ],
  "dat": [
    "luču",
    "lučam",
  ],
  "gen": [
    "luča",
    "lučev",
  ],
  "ins": [
    "lučem",
    "lučami",
  ],
  "loc": [
    "luču",
    "lučah",
  ],
  "nom": [
    "luč",
    "luče",
  ],
  "voc": [
    "luču",
    "luče",
  ],
}
`;

exports[`noun miscellaneous 498 (as feminine): feminine 1`] = `
{
  "acc": [
    "kųdeŕ",
    "kųderi",
  ],
  "dat": [
    "kųderi",
    "kųdeŕam",
  ],
  "gen": [
    "kųderi",
    "kųderij",
  ],
  "ins": [
    "kųdeŕjų",
    "kųdeŕami",
  ],
  "loc": [
    "kųderi",
    "kųdeŕah",
  ],
  "nom": [
    "kųdeŕ",
    "kųderi",
  ],
  "voc": [
    "kųderi",
    "kųderi",
  ],
}
`;

exports[`noun miscellaneous 498 (as masculine): masculine 1`] = `
{
  "acc": [
    "kųdeŕ",
    "kųdeŕe",
  ],
  "dat": [
    "kųdeŕu",
    "kųdeŕam",
  ],
  "gen": [
    "kųdeŕa",
    "kųdeŕev",
  ],
  "ins": [
    "kųdeŕem",
    "kųdeŕami",
  ],
  "loc": [
    "kųdeŕu",
    "kųdeŕah",
  ],
  "nom": [
    "kųdeŕ",
    "kųdeŕe",
  ],
  "voc": [
    "kųdeŕu",
    "kųdeŕe",
  ],
}
`;

exports[`noun miscellaneous 2155 (as feminine): feminine 1`] = `
{
  "acc": [
    "pųť",
    "pųti",
  ],
  "dat": [
    "pųti",
    "pųťam",
  ],
  "gen": [
    "pųti",
    "pųtij",
  ],
  "ins": [
    "pųťjų",
    "pųťami",
  ],
  "loc": [
    "pųti",
    "pųťah",
  ],
  "nom": [
    "pųť",
    "pųti",
  ],
  "voc": [
    "pųti",
    "pųti",
  ],
}
`;

exports[`noun miscellaneous 2155 (as masculine): masculine 1`] = `
{
  "acc": [
    "pųť",
    "pųťe",
  ],
  "dat": [
    "pųťu",
    "pųťam",
  ],
  "gen": [
    "pųťa",
    "pųťev",
  ],
  "ins": [
    "pųťem",
    "pųťami",
  ],
  "loc": [
    "pųťu",
    "pųťah",
  ],
  "nom": [
    "pųť",
    "pųťe",
  ],
  "voc": [
    "pųťu",
    "pųťe",
  ],
}
`;

exports[`noun miscellaneous 2715 (as feminine): feminine 1`] = `
{
  "acc": [
    "měď",
    "mědi",
  ],
  "dat": [
    "mědi",
    "měďam",
  ],
  "gen": [
    "mědi",
    "mědij",
  ],
  "ins": [
    "měďjų",
    "měďami",
  ],
  "loc": [
    "mědi",
    "měďah",
  ],
  "nom": [
    "měď",
    "mědi",
  ],
  "voc": [
    "mědi",
    "mědi",
  ],
}
`;

exports[`noun miscellaneous 2715 (as masculine): masculine 1`] = `
{
  "acc": [
    "měď",
    "měďe",
  ],
  "dat": [
    "měďu",
    "měďam",
  ],
  "gen": [
    "měďa",
    "měďev",
  ],
  "ins": [
    "měďem",
    "měďami",
  ],
  "loc": [
    "měďu",
    "měďah",
  ],
  "nom": [
    "měď",
    "měďe",
  ],
  "voc": [
    "měďu",
    "měďe",
  ],
}
`;

exports[`noun miscellaneous 3749 (as feminine): feminine 1`] = `
{
  "acc": [
    "uråvėnj",
    "uråvnji",
  ],
  "dat": [
    "uråvnji",
    "uråvnjam",
  ],
  "gen": [
    "uråvnji",
    "uråvnjij",
  ],
  "ins": [
    "uråvnjjų",
    "uråvnjami",
  ],
  "loc": [
    "uråvnji",
    "uråvnjah",
  ],
  "nom": [
    "uråvėnj",
    "uråvnji",
  ],
  "voc": [
    "uråvnji",
    "uråvnji",
  ],
}
`;

exports[`noun miscellaneous 3749 (as masculine): masculine 1`] = `
{
  "acc": [
    "uråvėnj",
    "uråvnje",
  ],
  "dat": [
    "uråvnju",
    "uråvnjam",
  ],
  "gen": [
    "uråvnja",
    "uråvnjev",
  ],
  "ins": [
    "uråvnjem",
    "uråvnjami",
  ],
  "loc": [
    "uråvnju",
    "uråvnjah",
  ],
  "nom": [
    "uråvėnj",
    "uråvnje",
  ],
  "voc": [
    "uråvnju",
    "uråvnje",
  ],
}
`;

exports[`noun miscellaneous 6181 (as feminine): feminine 1`] = `
{
  "acc": [
    "križ",
    "križi",
  ],
  "dat": [
    "križi",
    "križam",
  ],
  "gen": [
    "križi",
    "križij",
  ],
  "ins": [
    "križjų",
    "križami",
  ],
  "loc": [
    "križi",
    "križah",
  ],
  "nom": [
    "križ",
    "križi",
  ],
  "voc": [
    "križi",
    "križi",
  ],
}
`;

exports[`noun miscellaneous 6181 (as masculine): masculine 1`] = `
{
  "acc": [
    "križ",
    "križe",
  ],
  "dat": [
    "križu",
    "križam",
  ],
  "gen": [
    "križa",
    "križev",
  ],
  "ins": [
    "križem",
    "križami",
  ],
  "loc": [
    "križu",
    "križah",
  ],
  "nom": [
    "križ",
    "križe",
  ],
  "voc": [
    "križu",
    "križe",
  ],
}
`;

exports[`noun miscellaneous 15058 (as feminine): feminine 1`] = `
{
  "acc": [
    "svęź",
    "svęzi",
  ],
  "dat": [
    "svęzi",
    "svęźam",
  ],
  "gen": [
    "svęzi",
    "svęzij",
  ],
  "ins": [
    "svęźjų",
    "svęźami",
  ],
  "loc": [
    "svęzi",
    "svęźah",
  ],
  "nom": [
    "svęź",
    "svęzi",
  ],
  "voc": [
    "svęzi",
    "svęzi",
  ],
}
`;

exports[`noun miscellaneous 15058 (as masculine): masculine 1`] = `
{
  "acc": [
    "svęź",
    "svęźe",
  ],
  "dat": [
    "svęźu",
    "svęźam",
  ],
  "gen": [
    "svęźa",
    "svęźev",
  ],
  "ins": [
    "svęźem",
    "svęźami",
  ],
  "loc": [
    "svęźu",
    "svęźah",
  ],
  "nom": [
    "svęź",
    "svęźe",
  ],
  "voc": [
    "svęźu",
    "svęźe",
  ],
}
`;

exports[`noun miscellaneous 18959 (as feminine): feminine 1`] = `
{
  "acc": [
    "těnj",
    "těnji",
  ],
  "dat": [
    "těnji",
    "těnjam",
  ],
  "gen": [
    "těnji",
    "těnjij",
  ],
  "ins": [
    "těnjjų",
    "těnjami",
  ],
  "loc": [
    "těnji",
    "těnjah",
  ],
  "nom": [
    "těnj",
    "těnji",
  ],
  "voc": [
    "těnji",
    "těnji",
  ],
}
`;

exports[`noun miscellaneous 18959 (as masculine): masculine 1`] = `
{
  "acc": [
    "těnj",
    "těnje",
  ],
  "dat": [
    "těnju",
    "těnjam",
  ],
  "gen": [
    "těnja",
    "těnjev",
  ],
  "ins": [
    "těnjem",
    "těnjami",
  ],
  "loc": [
    "těnju",
    "těnjah",
  ],
  "nom": [
    "těnj",
    "těnje",
  ],
  "voc": [
    "těnju",
    "těnje",
  ],
}
`;

exports[`noun miscellaneous 23278 (as feminine): feminine 1`] = `
{
  "acc": [
    "pijanicų",
    "pijanice",
  ],
  "dat": [
    "pijanici",
    "pijanicam",
  ],
  "gen": [
    "pijanice",
    "pijanic",
  ],
  "ins": [
    "pijanicejų",
    "pijanicami",
  ],
  "loc": [
    "pijanici",
    "pijanicah",
  ],
  "nom": [
    "pijanica",
    "pijanice",
  ],
  "voc": [
    "pijanico",
    "pijanice",
  ],
}
`;

exports[`noun miscellaneous 23278 (as masculine): masculine 1`] = `
{
  "acc": [
    "pijanicų",
    "pijanice",
  ],
  "dat": [
    "pijanici",
    "pijanicam",
  ],
  "gen": [
    "pijanice",
    "pijanic",
  ],
  "ins": [
    "pijanicejų",
    "pijanicami",
  ],
  "loc": [
    "pijanici",
    "pijanicah",
  ],
  "nom": [
    "pijanica",
    "pijanice",
  ],
  "voc": [
    "pijanico",
    "pijanice",
  ],
}
`;

exports[`noun miscellaneous 31883 (as feminine): feminine 1`] = `
{
  "acc": [
    "sliź",
    "slizi",
  ],
  "dat": [
    "slizi",
    "sliźam",
  ],
  "gen": [
    "slizi",
    "slizij",
  ],
  "ins": [
    "sliźjų",
    "sliźami",
  ],
  "loc": [
    "slizi",
    "sliźah",
  ],
  "nom": [
    "sliź",
    "slizi",
  ],
  "voc": [
    "slizi",
    "slizi",
  ],
}
`;

exports[`noun miscellaneous 31883 (as masculine): masculine 1`] = `
{
  "acc": [
    "sliź",
    "sliźe",
  ],
  "dat": [
    "sliźu",
    "sliźam",
  ],
  "gen": [
    "sliźa",
    "sliźev",
  ],
  "ins": [
    "sliźem",
    "sliźami",
  ],
  "loc": [
    "sliźu",
    "sliźah",
  ],
  "nom": [
    "sliź",
    "sliźe",
  ],
  "voc": [
    "sliźu",
    "sliźe",
  ],
}
`;
